plugin.py

```diff
@@ -128,8 +197,168 @@ def run(self, edit: sublime.Edit) -> None:
+
+
+class RustAnalyzerMemoryUsage(LspTextCommand):
+    session_name = "rust-analyzer"
+
+    def run(self, edit: sublime.Edit) -> None:
+        session = self.session_by_name(self.session_name)
+        if session is None:
+            return
+        session.send_request(Request("rust-analyzer/memoryUsage"), self.on_result)
+
+    def on_result(self, payload: str) -> None:
+        window = self.view.window()
+        if window is None:
+            return
+        sheets = window.selected_sheets()
+        view = window.new_file(flags=sublime.TRANSIENT)
+        view.set_scratch(True)
+        view.set_name("--- RustAnalyzer Memory Usage ---")
+        view.run_command("append", {"characters": "Per-query memory usage:\n"})
+        view.run_command("append", {"characters": payload})
+        view.run_command("append", {"characters": "\n(note: database has been cleared)"})
+        view.set_read_only(True)
+        sheet = view.sheet()
+        if sheet is not None:
+            sheets.append(sheet)
+            window.select_sheets(sheets)
+
+
+RunnableArgs = TypedDict('RunnableArgs', {
+    'cargoArgs': List[str],
+    'cargoExtraArgs': List[str],
+    'executableArgs': List[str],
+    'overrideCargo': Optional[str],
+    'workspaceRoot': str,
+})
+Runnable = TypedDict('Runnable', {
+    'args': RunnableArgs,
+    'kind': str,
+    'label': str,
+})
+
+
+class RustAnalyzerExec(LspTextCommand):
+    session_name = "rust-analyzer"
+
+    def run(self, edit: sublime.Edit) -> None:
+        params = text_document_position_params(self.view, self.view.sel()[0].b)
+        session = self.session_by_name(self.session_name)
+        if session is None:
+            return
+        session.send_request(Request("experimental/runnables", params), self.on_result)
+
+    def run_terminus(self, check_phrase: str, payload: List[Runnable]) -> None:
+        window = self.view.window()
+        if window is None:
+            return
+
+        if len(payload) == 0:
+            return
+
+        output = None
+        for item in payload:
+            if item["label"].startswith(check_phrase):
+                output = item
+                break
+
+        if not output:
+            return
+
+        if not Terminus:
+            sublime.error_message(
+                'Cannot run executable "{}": You need to install the "Terminus" package and then restart Sublime Text'.format(output["kind"]))
+            return
+        if output["args"]["overrideCargo"]:
+            cargo_path = output["args"]["overrideCargo"]
+        else:
+            if not shutil.which("cargo"):
+                sublime.error_message('Cannot find "cargo" on path.')
+                return
+            cargo_path = '"{}"'.format(shutil.which("cargo"))
+        command_to_run = [cargo_path] + output["args"]["cargoArgs"] + \
+            output["args"]["cargoExtraArgs"] + output["args"]["executableArgs"]
```

Comment:
This whole logic here is very similar to what the `on_pre_server_command` does. It would be worth extracting that into a common function.

Also is `output["args"]["executableArgs"]` specific to this command? The other place doesn't have it.